libs/utils/src/properties.c

```diff
@@ -191,324 +481,449 @@ static void parseLine(const char* line, celix_properties_t *props) {
+    }
+
+    int rc = 0;
 
-    bool firstTime = true;
-    do {
-        if (firstTime){
-            line = strtok_r(in, "\n", &saveLinePointer);
-            firstTime = false;
-        }else {
-            line = strtok_r(NULL, "\n", &saveLinePointer);
+    if (header && strstr("\n", header)) {
+        celix_err_push("Header cannot contain newlines. Ignoring header.");
+    } else if (header) {
+        rc = fputc('#', file);
+        if (rc != 0) {
+            rc = fputs(header, file);
+        }
+        if (rc != 0) {
+            rc = fputc('\n', file);
         }
+    }
 
-        if (line == NULL){
-            break;
+    CELIX_PROPERTIES_ITERATE(properties, iter) {
+        const char* val = iter.entry.value;
+        if (rc != EOF) {
+            rc = celix_properties_storeEscapedString(file, iter.key);
+        }
+        if (rc != EOF) {
+            rc = fputc('=', file);
+        }
+        if (rc != EOF) {
+            rc = celix_properties_storeEscapedString(file, val);
+        }
+        if (rc != EOF) {
+            rc = fputc('\n', file);
         }
+    }
+    if (rc != EOF) {
+        rc = fclose(file);
+    } else {
+        fclose(file);
+    }
+    if (rc == EOF) {
+        celix_err_push("Failed to write properties to file");
+        return CELIX_FILE_IO_EXCEPTION;
+    }
+    return CELIX_SUCCESS;
+}
 
-        parseLine(line, props);
-    } while(line != NULL);
+celix_properties_t* celix_properties_copy(const celix_properties_t* properties) {
+    celix_properties_t* copy = celix_properties_create();
 
-    free(in);
+    if (!copy) {
+        celix_err_push("Failed to create properties copy");
+        return NULL;
+    }
 
-    return props;
+    if (!properties) {
+        return copy;
+    }
+
+    CELIX_PROPERTIES_ITERATE(properties, iter) {
+        celix_status_t status;
+        if (iter.entry.valueType == CELIX_PROPERTIES_VALUE_TYPE_STRING) {
+            status = celix_properties_set(copy, iter.key, iter.entry.value);
+        } else if (iter.entry.valueType == CELIX_PROPERTIES_VALUE_TYPE_LONG) {
+            status = celix_properties_setLong(copy, iter.key, iter.entry.typed.longValue);
+        } else if (iter.entry.valueType == CELIX_PROPERTIES_VALUE_TYPE_DOUBLE) {
+            status = celix_properties_setDouble(copy, iter.key, iter.entry.typed.doubleValue);
+        } else if (iter.entry.valueType == CELIX_PROPERTIES_VALUE_TYPE_BOOL) {
+            status = celix_properties_setBool(copy, iter.key, iter.entry.typed.boolValue);
+        } else /*version*/ {
+            assert(iter.entry.valueType == CELIX_PROPERTIES_VALUE_TYPE_VERSION);
+            status = celix_properties_setVersion(copy, iter.key, iter.entry.typed.versionValue);
+        }
+        if (status != CELIX_SUCCESS) {
+            celix_err_pushf("Failed to copy property %s", iter.key);
+            celix_properties_destroy(copy);
+            return NULL;
+        }
+    }
+    return celix_steal_ptr(copy);
```

Comment:
```suggestion
    return copy;
```
copy is not an auto variable.